### PR TITLE
collapse/expand all collections

### LINKED ISSUE #22781: Ability to expand/collapse all collections in data model
### Describe the Improvement

I really like the expand/collapse feature for nested collections in the data model, but for big projects, it would be great to improve that to make an expand/collapse button for all collections at once, rather than one at a time.

### PR DESCRIPTION
**Pull request by drennvinn**
I'm not sure if I did it the cleanest way possible. Feel free to correct me.

https://github.com/directus/directus/assets/18407632/e60328da-05ef-4d4b-a6bf-cc777eb6123f

Fixes [#22781](https://github.com/directus/directus/issues/22781)

### CODE REVIEW

**Comment by rijkvanzanten** [comment]
I'm wondering if the header bar is the appropriate place to render this 🤔 That's usually reserved for actions that apply module wide[^1] and I'm not sure if the Icon is clear enough.

Maybe we should go with a more traditional "Expand **All** / None" written text buttons on the top right of the list. I'm gonna noodle on some design directions and circle back with final thoughts! 

[^1]: .. though we do have full-screen as part of Insights as well

File: app/src/modules/settings/routes/data-model/collections/collections.vue [hunk 1]
```diff
@@ -23,6 +24,12 @@ const editCollection = ref<Collection | null>();
 
 const collectionsStore = useCollectionsStore();
 
+const {data: isCollectionsExpanded} = useLocalStorage('settings-collapsed-all', true as boolean);
```
Comment (rijkvanzanten) [inline app/src/modules/settings/routes/data-model/collections/collections.vue:27]: Keeping this in localStorage is a nice touch :)

**Comment by drennvinn** [comment]
> I'm wondering if the header bar is the appropriate place to render this 🤔 

Maybe it's time to introduce a new slot in private-view component like secondary header or menu? 😅
I faced this specific action button location issue several times when creating custom modules.

Capture d’écran 2024-06-18 à 19 33 20

**Comment by rijkvanzanten** [comment]
Ha not a bad idea! Maybe a bit scope-creepy for this particular PR, but in general I can see that happening! I've actually been dreaming of a more "slot-based" extension type where we just define a bunch of slots, and the extension can communicate which ones it wants to be rendered in.

For this particular PR, lets go with an inline-text button "Expand All / None" where all and none are clickable buttons (probably in primary) at the top right of the list 🙂

**Comment by drennvinn** [comment]
@rijkvanzanten I did this based on what you suggested. I tried a combination of outlined and non-outlined buttons because I thought it looked nice. Could this be suitable?

https://github.com/directus/directus/assets/18407632/1957fcfb-4113-4725-92ed-d9adc7a3a04e

**Comment by rijkvanzanten** [comment]
@drennvinn I think the double toggle has the problem that it doesn't have an indeterminate state. You will have some collapsed some sections but not all. I think we can keep it simple and literally just have two "links" _Expand [all] / [none]_ that collapses all or none respectively. There doesn't have to be an active state, as it's more like a one-off action to collapse all, rather than a permanent state 🤔

**Comment by drennvinn** [comment]
Here is an initial draft regarding the addition of the Expand All/None buttons. However, I have encountered difficulties with managing and saving states after page reloads. There are multiple contexts to manage (line 43 in collection-item.vue). I would appreciate a code review and advice on how to improve this (if necessary).
Feel free to suggest any changes if needed.


https://github.com/directus/directus/assets/18407632/6bf1b67f-0b9b-4018-b83c-2bcaf74ec2d2

File: app/src/modules/settings/routes/data-model/collections/components/collection-item.vue [hunk 2]
```diff
@@ -10,23 +10,69 @@ import { CollectionTree } from '../collections.vue';
...
+				} else {
+					isCollectionExpanded.value = expanded;
+				}
+
+				break;
+		}
+	},
+	{ immediate: true },
+);
+
```
Comment (drennvinn) [inline app/src/modules/settings/routes/data-model/collections/components/collection-item.vue:75]: This approach is not very clean 😅, but I wanted to first confirm if this meets your expectations before refactoring.

**Comment by rijkvanzanten** [comment]
@drennvinn Instead of saving the collapsed state on a "all" or "nothing" basis, I think we have to save it on a per-row basis. That way we can persist the exact combination of open/closed rows rather than having to force everything open or shut 🤔

**Comment by rijkvanzanten** [comment]
The way I'd imagine that to work is to treat everything as open by default, and then store an array of keys of the items that should be collapsed. That way each row can check against that array and the array can be synced with localstorage. The "Expand All" can then set that array to `[]`, and the "Collapse All" sets the array to every row ID that exists on the page.

File: app/src/modules/settings/routes/data-model/collections/components/collection-item.vue (hunk 2 above)
Comment (rijkvanzanten) [reply app/src/modules/settings/routes/data-model/collections/components/collection-item.vue:75]: I have an updated strategy below 🙂 

Other than that, an `if/else` statement is probably better suited than a `switch` for a simple yes/no check

**Comment by drennvinn** [comment]
Thank you, @rijkvanzanten, for your suggestion. I also find it much more relevant. Here's what I've done:


https://github.com/directus/directus/assets/18407632/c4da427d-e3b9-4aa5-94db-16638aecab69


As you can see, I've stored two values in localstorage: `collapsedIds` and `expandedIds`. These values are updated based on user interactions. This approach helps maintain the state for all collections, whether expanded or collapsed, even after a page reload.

**Comment by rijkvanzanten** [comment]
Almost there @drennvinn! Seeing that the collapsed/open state is always a boolean (it can never be open and closed at the same time), there's no need to have two arrays. By definition, if the item is not in the "Closed" array, it's supposed to be opened! That should simplify the code quite a bit 🙂

**Comment by drennvinn** [comment]
You are absolutely right @rijkvanzanten ! and sorry, I only have very little time to devote to this PR during my days, and so I forget the most important part, which is to step back and reflect on what I have just done. 😅
I removed the `expandedIds` array, so I think it's ready for review now.

**Comment by rijkvanzanten** [comment]
All good and no hard feelings! I'll push one more commit with some opinionated code style changes and we'll get this thing merged 🚀 Thanks for your contribution!

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**